### ISSUE TITLE
Log prob computation in TDMPC2

### ISSUE THREAD

**Issue by wertyuilife2** [nicklashansen/tdmpc2#44]
I have some questions about the way logprob is calculated in TDMPC2.

The code below shows that the logprob result in TDMPC2 seems to differ from `torchrl` and SAC.

```python
import torch
import torch.nn.functional as F
from torchrl.modules.distributions import TanhNormal

@torch.jit.script
def _gaussian_residual(eps, log_std):
	return -0.5 * eps.pow(2) - log_std

@torch.jit.script
def _gaussian_logprob(residual):
	return residual - 0.5 * torch.log(2 * torch.pi)

def gaussian_logprob(eps, log_std, size=None):
	"""
	origin tdmpc2 code
	multiply action dim size on logprob
	and first sum, then minus constant number.
	"""
	residual = _gaussian_residual(eps, log_std).sum(-1, keepdim=True)
	if size is None:
		size = eps.size(-1)
	return _gaussian_logprob(residual) * size

def gaussian_logprob_sac(eps, log_std):
	"""SAC gaussian logprob code"""
	residual = _gaussian_residual(eps, log_std)
	return _gaussian_logprob(residual).sum(-1, keepdim=True)	

@torch.jit.script
def _squash(pi):
	return torch.log(F.relu(1 - pi.pow(2)) + 1e-6)

def squash(mu, pi, log_pi): 
	"""Apply squashing function."""
	mu = torch.tanh(mu)
	pi = torch.tanh(pi)
	log_pi -= _squash(pi).sum(-1, keepdim=True)
	return mu, pi, log_pi

if __name__ == "__main__":
	mu = torch.tensor([[-0.2,-0.1],[0.1,0.2]],dtype=torch.float)
	log_std = torch.tensor([[-2,-1],[0,1]],dtype=torch.float)
	dist = TanhNormal(loc=mu, scale=log_std.exp())	
	
	eps = torch.tensor([[0.1,0.2],[0.3,0.4]],dtype=torch.float)
	pi = mu + eps * log_std.exp()
 
	# compute logprob in sac
	log_pi = gaussian_logprob_sac(eps, log_std)
	_, _, log_pi = squash(mu, pi, log_pi)
	print("sac logprob:", log_pi)
 
 	# compute logprob in tdmpc
	log_pi = gaussian_logprob(eps, log_std, size=None)
	mu, pi, log_pi = squash(mu, pi, log_pi)
	print("tdmpc logprob:", log_pi)

	# compute logprob using torchrl
	logprob = dist.log_prob(pi)
	print("torchrl logprob:", logprob)

```
code result:
```
sac logprob: tensor([[ 1.1724],
        [-1.4718]])
tdmpc logprob: tensor([[ 4.1474],
        [-2.5968]])
torchrl logprob: tensor([ 1.1724, -1.4718])
```

Specifically, TDMPC2 multiplies the logprob result by the action dimension during calculation. Could you please explain why this is done?

**Comment by nicklashansen** [comment]
Thanks for pointing out this discrepancy! The original motivation for this implementation was to account for masked out action dimensions in the multi-task case, but I agree that should ideally be addressed in a way that does not alter the log probabilities when no masking is applied. I'll try to issue a fix soon. Either way, I don't believe this discrepancy should affect results in any meaningful way.

**Comment by wertyuilife2** [comment]
Thanks for the explanation! I also agree that this isn’t a fundamental issue—I was just genuinely curious about the reasoning behind this design choice.

However, based on my previous experiments, at least on the Dog-Run task in DMControl, multiplying by the action dim does seem to bring some benefits compared to not doing so. I suggest paying attention to the performance on this task when applying the fix.

**Comment by nicklashansen** [comment]
Interesting! It's possible that the scaling increases policy entropy a bit for tasks with large action spaces? I wonder if scaling the entropy coef with action space dim would yield the same result.

**Comment by wertyuilife2** [comment]
Scaling the entropy coef and scaling the log prob don't seem mathematically identical due to the additional squash() step, but I believe they will have similar effects.

**Comment by ShaneFlandermeyer** [comment]
And because @nicklashansen wanted to test this, here's the training reward for the original implementation (gray) and scaling the entropy coefficient directly (blue):

![Screenshot from 2024-11-28 09-09-18](https://github.com/user-attachments/assets/9a9ce4ab-41d5-49be-a2cc-e298cda63149)

**Comment by nicklashansen** [comment]
Interesting, thanks for running this! I wonder if this is more of a one-off phenomenon for Dog or if it applies to high-dimensional action spaces in general. I would caution against making conclusions based off on a small set of experiments since it's still RL after all. I'll see if I can run some experiments on this soon :-)

**Comment by nicklashansen** [comment]
Ran some numbers on this. Tried 5 tasks of varying action dims: `dog-run, humanoid-run, quadruped-run, cheetah-run, walker-run`, 2 seeds each. Experimented with 4 implementations:

- **default**: current TD-MPC2 implementation with default hparams (roughly equivalent to a non-linear scaling of entropy based on action dim)
- **sac-logprob**: the SAC implementation as implemented by @wertyuilife2
- **sac-logprob-1e-3**: same SAC implementation but with a 10x larger entropy coefficient
- **scale-post-squash**: same SAC implementation but with log prob scaled by action dim *after* squashing (equivalent to scaling the entropy coefficient by action dim)

**dog-run** (36-d)
<img width="300" alt="image" src="https://github.com/user-attachments/assets/02dc3314-bc53-4654-888e-a42271162668">

**humanoid-run** (24-d)
<img width="300" alt="image" src="https://github.com/user-attachments/assets/ee19236c-0881-4f21-a52e-d7c3d2ef5369">

**quadruped-run** (12-d)
<img width="299" alt="image" src="https://github.com/user-attachments/assets/ca31b072-ad76-4422-8d4b-b7053b2d1701">

Skipping `cheetah-run, walker-run` (6-d) since results are the same for all implementations on these tasks.

It seems to me that this is mostly a matter of entropy coefficient, and that it affects the `dog-run` task the most. The reasons for this would be interesting to investigate further, e.g., is this specific to the Dog tasks of DMControl or would the effect be even more pronounced with a task with say 100-d action space?

**Comment by ShaneFlandermeyer** [comment]
Those results make sense to me. Would it make more sense for the policy function to separately return the "true" log probabilities and the action space-dependent scale factor? That way the math is correct but you still get the performance benefits. The scale factor would just be the ratio between the true log probs and the ones calculated in the current code.

Something like this pseudocode which I obviously haven't tested yet:

```
    # Sample action and compute logprobs
    eps = jax.random.normal(key, mu.shape)
    true_log_probs = jax.scipy.stats.norm.logpdf(x_t, loc=mu, scale=std).sum(-1)
    x_t = mu + eps * std

    # Squash tanh
    mean = jnp.tanh(mu)
    action = jnp.tanh(x_t)
    true_log_probs -= jnp.log(1 - action**2 + 1e-6).sum(-1)

    # Compute entropy scale proportional to action dimension
    # Identical to the original implementation
    residual = (-0.5 * eps**2 - log_std).sum(-1)
    scaled_log_probs = x_t.shape[-1] * (residual - 0.5 * jnp.log(2 * jnp.pi))
    scaled_log_probs -= jnp.log(nn.relu(1 - action**2) + 1e-6).sum(-1)
    entropy_scale = scaled_log_probs / true_log_probs
    return action, mean, log_std, true_log_probs, entropy_scale
```

**Comment by nicklashansen** [comment]
Yes, I think that's a very reasonable solution! I'm happy to push these changes. My implementation might differ slightly since I want to future-proof it wrt. other planned features :-)

**Comment by ShaneFlandermeyer** [comment]
@wertyuilife2 This is a very interesting observation. Is it possible that the excessively large actions cause the agent to heavily "fall back" to planning during early training in large action spaces? I wonder if the logstd stays at the maximum the whole time for longer runs...

**Comment by wertyuilife2** [comment]
In fact, the network policy of TD-MPC2 does not perform well on tasks with high-dimensional action spaces. I have done some work to improve this. 

However, I am still unsure whether the weakness of the network policy is related to this inappropriate entropy scale. I tried modifying the entropy calculation to address this problem, but so far, the results have been strange, and there hasn’t been any meaningful progress :(

**Comment by ShaneFlandermeyer** [comment]
I also don't expect the log-std to ever go below 2. I'll do some tests in dog-run which vary the number of policy prior samples. It's possible that it's weighting the policy prior too heavily (or not at all).

**Comment by wertyuilife2** [comment]
> Due to the excessively large std, the distribution is almost unaffected by the mean.

> As a result, the output actions are essentially -1 for 50% of the time and 1 for the other 50%.

Oh, I might be wrong. The mean is derived from the output of the neural network without activation, so its value may be comparable to the standard deviation. Therefore, the distribution between -1 and 1 may not necessarily be evenly split; instead, the ratio is determined by the value of the mean. However, overall, the values should still primarily be either -1 or 1.

**Comment by nicklashansen** [comment]
It is possible that extreme values make value estimation easier when actions are high-dim?

**Comment by wertyuilife2** [comment]
> It is possible that extreme values make value estimation easier when actions are high-dim?

It makes some sense to me, but it still doesn’t convince me of its validity.

> In fact, the network policy of TD-MPC2 does not perform well on tasks with high-dimensional action spaces.

I somewhat believe that this extreme value has a negative influence on the max-Q gradient for the prior policy. More specifically, it might cause gradient vanishing under the tanh function. I have some results, but it might be more appropriate to share them later since they are part of a paper submission.

**Comment by ShaneFlandermeyer** [comment]
I'm also very interested in your results. I don't think this is a huge discrepancy, but it's still bothering me more than it probably should!

**Comment by nicklashansen** [comment]
Commit https://github.com/nicklashansen/tdmpc2/commit/32fc2bdf934004ac2e87ff60debdfd186ae9ed15 changes the policy to now output the sampled action along with a dictionary containing other relevant outputs, including log probabilities and a coefficient for the entropy loss.

**Comment by nicklashansen** [comment]
I will close this issue since it has mostly been resolved, but please feel free to reopen if you have any additional thoughts/questions on the topic!

**Comment by ShaneFlandermeyer** [comment]
Out of curiosity, have either of y'all come up with a more intuitive explanation for this behavior @nicklashansen @wertyuilife2?

**Comment by wertyuilife2** [comment]
Oh, for sure! I actually noticed this issue (i.e., TDMPC2's prior policy underperforming) while working on my ICLR 2025 submission, which got accepted. Happy to share some insights!

The results are a bit detailed, so I’d suggest checking out the curves in the [paper](https://openreview.net/forum?id=i7jAYFYDcM&noteId=i6zMv7RzgX) (Figure 2 and Figure 10 show the issue).

In short, I applied the idea of expert iteration within the TDMPC2 framework, creating an AlphaZero-like algorithm called BMPC. This avoids the failure of prior policy learning and improves performance on some tough tasks.

That said, I haven’t explored the root cause of the issue in depth. My intuition is that it’s due to the lack of robustness in gradient-based optimization (like DDPG-style max-Q gradients) for high-dimensional problems. I’ve also discussed this briefly in my OpenReview response to Reviewer tbYo’s Question 1:

>This is a crucial question! On the surface, we tend to think of it this way:
>
>(1) TD-MPC's network policy is trained similarly to SAC, with only two differences: The training data comes from the MPC policy, > which is of higher quality; The inputs are well-represented latent vectors rather than raw observations.
>
>(2) SAC tends to fail on high-dimensional tasks.
>
>(3) Hypothesis: Even with better data and better representations, SAC still fails.
>
>(4) Therefore, TD-MPC's network policy also fails.
>
>As for why SAC fails, one potential reason is that for high action-dim tasks, sample-based optimization (like MPPI) is more robust >and suitable than gradient-based optimization (like DDPG-style max-Q gradient).
>
>In fact, we tried to combine both expert imitation and max-Q gradient for policy optimization, and it showed fairly good results. However, we have not yet explored this approach in depth. We believe there are many more things to think about, and more experiments are needed to uncover the underlying reasons behind the phenomenon.

As I said in the issue, I also believe that the extreme std value has a negative influence on the max-Q gradient for the prior policy. However, as shown in the experimental results in the issue, TDMPC2's performance relies on these large std values, so I'm not sure how to address this either.

BMPC is simple to implement, and I’ve already released an initial anonymous open-source version. I’m planning to rewrite it on the latest accelerated TDMPC2 codebase for better decoupling. January was a bit busy, so I haven’t finished yet, but I’ll get to it soon. If @nicklashansen  is interested, I could even add it as a branch or feature in the TDMPC2 repo.

**Comment by ShaneFlandermeyer** [comment]
Very interesting, thanks for the insight! I'll have to check out the paper and code for sure. I would definitely be interested in adding this to the Jax repo

**Comment by ShaneFlandermeyer** [comment]
@wertyuilife2 Just read through the paper, and I really like your approach!! I'll try to Jax-ify things in the next day or two and reach out if I have any questions.

**Comment by wertyuilife2** [comment]
@ShaneFlandermeyer Glad you liked it! Feel free to reach out with any questions.

**Comment by ShaneFlandermeyer** [comment]
@wertyuilife2  If I'm reading things correctly, it looks like BMPC uses the same scaling on the logprobs as TDMPC2? Have you compared that to the "correct" scaling computed from the gaussian_logprob_sac function in the first comment in this thread? Apologies if you answered this in your first comment today and I'm just misunderstanding

**Comment by wertyuilife2** [comment]
@ShaneFlandermeyer Yes, to keep things fair, I also scaled the entropy loss by the action dimension in BMPC.

That said, scaling the entropy loss doesn’t make much difference for BMPC. Even with the entropy loss coef set to 0, it performs almost the same on tasks like dog-run and humanoid-run. But for exploration-heavy tasks like cartpole-swingup-sparse, a low policy std makes it hard to find the first reward.

Also, because of changes in the policy loss, BMPC’s log-std doesn’t get as high as 2. It usually stays in a more reasonable range, like around 0.3.

**Comment by ShaneFlandermeyer** [comment]
@wertyuilife2  Ok, I've implemented everything in Jax but the agent struggles to learn much of anything in dog-run (doesn't really exceed 200 reward). Would you be willing to take a look through the code and see if you can spot any problems? It's in Jax, but everything matches this repo closely and should be readable. I also propose we move the discussion to [here](https://github.com/ShaneFlandermeyer/tdmpc2-jax/issues/13) since this is unrelated to the logprob stuff.

**Comment by nicklashansen** [comment]
@wertyuilife2 That's very interesting, thanks for sharing! I will take a closer look at the paper soon. Congrats on the acceptance!

**Comment by niamorg** [comment]
Hi,

Thanks @nicklashansen for pinpointing this discussion (coming from issue https://github.com/nicklashansen/tdmpc2/issues/76), which somewhat answers my question since I think I understand now why the entropy term is what it is.

Sorry to bring this up again, but I still wanted to point several things (I may be wrong but I do not think so):

1) In the original TD-MPC2 implementation, the computed log probabilities were not the exact SAC log_probs. From what I saw, and as it was described at the top of this thread, they were

$$\displaystyle \text{logprobs}_\text{TDMPC2} = d \times \left[\sum_i -\frac12 \varepsilon_i^2 - \log(\sigma_i)\right] - \log\left((\sqrt{2\pi})^d\right) - \sum_i \log(1 - a_i^2)$$

instead of
 
$$\displaystyle \text{logprobs}_\text{SAC} = 1 \times \left[\sum_i -\frac12 \varepsilon_i^2 - \log(\sigma_i)\right] - \log\left((\sqrt{2\pi})^d\right)  - \sum_i \log(1 - a_i^2)$$

where $d$ is the action dim. The difference being that the "gaussian residual" part is scaled by $d$ (which I am not able to interpret).

The entropy term was just $-\text{coeff}_{\text entropy} \times  \text{logprobs}$ (averaged over the batch).

2) In the current implementation, the entropy term seems to be:

$$\displaystyle -\text{coeff}\_{\text entropy} \times d \times \left[  -\sum_i \frac12 \varepsilon_i^2 - \log(\sigma_i) -\log\left((\sqrt{2\pi})^d\right) \right] =-\text{coeff}_\text{entropy}  \times (d \times \text{logprobs}\_\text{gaussian})$$

Therefore the current entropy term is proportional to the entropy of the **non** squashed gaussian, not the squashed gaussian (from what I can see in `WorldModel.pi()` and `TDMPC2.update_pi()`). I am not sure if this is intentional. I think the current implementation is a bit inspired by the above JAX code of @ShaneFlandermeyer (https://github.com/nicklashansen/tdmpc2/issues/44#issuecomment-2510268922) which is equivalent to the original TDMPC2 implementation. However the current version doesn't follow exactly the JAX code since (first) the `scaled_log_probs` variable lacks the "squashed log probs" term and (second) is computed from $d \times \text{logprobs}\_\text{gaussian}$ instead of the "inexact" formula used in the original implementation (where the constant term $\log\left((\sqrt{2\pi})^d\right)$ is **not** multiplied by $d$).

I do not know if you think thats a problem or if it is intentional but I prefered pointing it out.

**Comment by nicklashansen** [comment]
Thanks @niamorg for bringing this to my attention! I will need to look more into this but I indeed intended for the updated pytorch implementation to match the original. If that is not the case then it definitely should be fixed :-) That said, I suspect that any performance difference should be relatively small.

**Comment by niamorg** [comment]
A quick look will convince you that the "squashed_logprobs" term is missing (the correctional term to add to the "gaussian_logprobs" to get the p.d.f. of the squashed gaussian). Also, the reason the "inexact" formula for the "gaussian_logprobs" is not used anymore is because the `math.gaussian_logprob()` function has been corrected!
